Simplicity-TR.tm

```diff
@@ -2298,6 +2298,34 @@
   Since the <math|SHA256<rsub|<2><rsup|8>>> variant is so commonly used, we
   will write it unadorned as simply <math|SHA256>.
 
+  <subsubsection|Tagged Hashes>
+
+  BIP-340, BIP-341 used a ``tagged-hash'' format to help avoid message
+  collisions between different message variants. We adopt the same format
+  here to avoid message collisions within a Simplicity application and we
+  hope to avoid message collision across various applications.
+
+  The tagged hash format begins every message with one block prefix which
+  indicate the variant, or type of digest data being hashed. This first block
+  consists of the SHA-256 digest of a string repeated twice. \ Given a byte
+  string <math|t\<of\><around*|(|<2><rsup|8>|)><rsup|\<ast\>>>, we can
+  defined the tagged hash of bit strings,
+  <math|SHA256<rsup|t><rsub|<2>>\<of\><2><rsup|\<ast\>>\<rightarrow\><2><rsup|256>>,
+  and byte strings, \ <math|SHA256<rsup|t><rsub|<2><rsup|8>>\<of\><around*|(|<2><rsup|8>|)><rsup|\<ast\>>\<rightarrow\><2><rsup|256>>.
+
+  <\eqnarray*>
+    <tformat|<table|<row|<cell|SHA256<rsup|t><rsub|<2>><around*|(|x|)>>|<cell|\<assign\>>|<cell|SHA256<rsub|<2>><around*|(|\<mu\><rsup|\<ast\>><around*|(|<around*|(|\<iota\><rsup|<2><rsup|8>><rsub|<2><rsup|\<ast\>>>|)><rsup|\<ast\>><around*|(|p\<cdummy\>p|)>|)>\<cdummy\>x|)>>>|<row|<cell|SHA256<rsup|t><rsub|<2><rsup|8>><around*|(|x|)>>|<cell|\<assign\>>|<cell|SHA256<rsub|<2><rsup|8>><around*|(|p\<cdummy\>p\<cdummy\>x|)>>>|<row|<cell|>|<cell|where>|<cell|p\<assign\>BE<rsub|256><around*|(|SHA256<rsub|<2><rsup|8>><around*|(|t|)>|)>>>>>
+  </eqnarray*>
+
+  \ We can compute the SHA-256 midstate after compressing this first block
+  containing the prefix tag.
+
+  <\equation*>
+    SHA256<rsup|t><rsub|IV>\<assign\>SHA256<rsub|Block><around*|\<langle\>|SHA256<rsub|IV>,\<Delta\><around*|(|SHA256<rsub|<2><rsup|8>><around*|(|t|)>|)>|\<rangle\>>
```

Comment:
Should there be another SHA256 before the \Delta pair?

Reply:
`SHA256[Block]` has type `2^256 * 2^512 -> 2^256`.  `SHA256[2^8](t)` has type `2^256`.  The Delta function duplicates its argument so  `\Delta(SHA256[2^8](t))` has type `2^256 * 2^256` which is equal to `2^512`.  This is the type required at that position.

Reply:
Ah, I thought `\Delta` infix operator for pair. So, in case of a single argument, It duplicates it arg.

Reply:
Right.  The `\Delta` function creates a pair of duplicates while the infix `white triangle` operator pairs up the results of two functions.  I can see how the notation might be confusing, although it is (somewhat) standard.

`\Delta := [[iden \whitetriangle iden]]`